drivers/input/input_tsc_keys.c

```diff
@@ -0,0 +1,439 @@
+/*
+ * Copyright (c) 2024 Arif Balik <arifbalik@outlook.com>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <soc.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <autoconf.h>
+#include <zephyr/irq.h>
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/input/input.h>
+#include <zephyr/drivers/reset.h>
+#include <zephyr/sys/ring_buffer.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+
+LOG_MODULE_REGISTER(tsc_keys, CONFIG_INPUT_LOG_LEVEL);
+
+#define DT_DRV_COMPAT st_stm32_tsc
+
+/* each group only has 4 configurable I/O */
+#define GET_GROUP_BITS(val) (uint32_t)(((val) & 0x0f) << ((group->group - 1) * 4))
+
+typedef void (*stm32_tsc_callback_t)(uint8_t group, uint32_t count, void *user_data);
+
+struct stm32_tsc_group_config {
+	uint8_t group;
+	uint8_t channel_ios;
+	uint8_t sampling_io;
+	bool use_as_shield;
+};
+
+struct stm32_tsc_config {
+	const TSC_TypeDef *tsc;
+	const struct stm32_pclken *pclken;
+	struct reset_dt_spec reset;
+	const struct pinctrl_dev_config *pcfg;
+	const struct stm32_tsc_group_config *groups;
+	uint8_t group_cnt;
+
+	uint32_t pgpsc;
+	uint8_t ctph;
+	uint8_t ctpl;
+	bool spread_spectrum;
+	uint8_t sscpsc;
+	uint8_t ssd;
+	uint16_t max_count;
+	bool iodef;
+	bool sync_acq;
+	bool sync_pol;
+	void (*irq_func)(void);
+};
+
+struct stm32_tsc_data {
+	stm32_tsc_callback_t cb;
+	void *user_data;
+};
+
+void stm32_tsc_start(const struct device *dev)
+{
+	const struct stm32_tsc_config *config = dev->config;
+
+	/* clear interrupts */
+	sys_set_bits((mem_addr_t)&config->tsc->ICR, TSC_ICR_EOAIC | TSC_ICR_MCEIC);
+
+	/* enable end of acquisition and max count error interrupts */
+	sys_set_bits((mem_addr_t)&config->tsc->IER, TSC_IER_EOAIE | TSC_IER_MCEIE);
+
+	/* TODO: When sync acqusition mode is enabled, both this bit and an external input signal
+	 * should be set. When the acqusition stops this bit is cleared, so even if a sync signal is
+	 * present, the next acqusition will not start until this bit is set again.
+	 */
+	/* start acquisition */
+	sys_set_bit((mem_addr_t)&config->tsc->CR, TSC_CR_START_Pos);
+}
+
+void stm32_tsc_register_callback(const struct device *dev, stm32_tsc_callback_t cb, void *user_data)
+{
+	struct stm32_tsc_data *data = dev->data;
+
+	data->cb = cb;
+	data->user_data = user_data;
+}
```

Comment:
mmh ok but does it still make sense to have this internal callback interface?

Reply:
Ok I see that now, I think it's quite confusing though, not sure that's what Gerard was meant to suggest, I don't mind the whole code living here and the bindings either meged or split but if there's two devices with an API in between I think it makes sense to have them in two files, so the previous layout made more sense to me.

@gmarull can you take a look again and see if this is what you were asking for?

Reply:
Yeah, i think there was definitely a miscommunication, thats why i made this half completed change, so that we can take a look. I kept the clone of the old one so it should be no problem to revert it back

Reply:
Yes, correct. Altough I haven't seen one, the app notes and the peripheral documents suggest many use cases

Reply:
ok then I think the previous one made sense, wait for Gerard to take a look though